search/saddleback_search.cpp

```diff
@@ -0,0 +1,106 @@
+#include <vector>       /// for std::vector
+
+/** \namespace search
+ * \brief Algorithms for searching
+ */
+namespace search {
+/** \namespace saddleback
+ * \brief Function for implementing [Saddleback Algorithm](https://www.geeksforgeeks.org/saddleback-search-algorithm-in-a-2d-array).
+ */
+namespace saddleback {
+/**
+ * This function implements [Saddleback Algorithm](https://www.geeksforgeeks.org/saddleback-search-algorithm-in-a-2d-array),
+ * on a sorted 2D array, and finds the location of the element needed to search
+ * @param matrix 2D matrix which is sorted on the basis of rows and columns
+ * @param element the element which is needed to be searched
+ * @return An std::pair of with row and column populated within it, if the
+ * element is present.
+ * @return An std::pair with (0, 0), if the element is not present.
+ */
+std::pair<uint64_t, uint64_t> saddleback(std::vector<std::vector<int32_t>> matrix,
+                               int32_t element) {
+    uint64_t left_index = 0;
+    uint64_t right_index = matrix[0].size() - 1;  // Start from top right corner
+    while (left_index < matrix.size()) {  // Exit once the value of indexes get out of range.
+        if (element ==
+            matrix[left_index]
+                  [right_index]) {  // If value on this position of matrix is
+                                    // equal to element, return (row, column).
+            return std::pair<uint64_t, uint64_t>(left_index+1, right_index+1);
+        } else if (element >
+                   matrix[left_index]
+                         [right_index]) {  // Else if value on this position of
+                                           // matrix is less than the element,
+                                           // move left.
+            ++left_index;
+        } else if (element <
+                   matrix[left_index]
+                         [right_index]) {  // Else if value on this position of
+                                           // matrix is greater than the
+                                           // element, move down.
+            if(!right_index)
+                break;
+            else --right_index;
+        }
+    }
+    return std::pair<uint64_t, uint64_t>(
+        0, 0);  // If the program reaches here, that means one of the index
+                  // went out of index, hence no element present.
+}
+}  // namespace saddleback
+}  // namespace search
+
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+static void test() {
+    std::vector<std::vector<int32_t>> matrix = {{1, 10, 100, 1000, 10000},
+                                            {2, 20, 200, 2000, 20000},
+                                            {3, 30, 300, 3000, 30000},
+                                            {4, 40, 400, 4000, 40000},
+                                            {5, 50, 500, 5000, 50000}};
+
+    std::pair<uint64_t, uint64_t> not_found = std::pair<uint64_t, uint64_t>(0, 0);
```

Comment:
Change all these instances (and below) to `uint32_t`.

Reply:
Okay